README.md
# Prettier 4GL Plugin

<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->

[![All Contributors](https://img.shields.io/badge/all_contributors-1-orange.svg?style=flat-square)](#contributors-)

<!-- ALL-CONTRIBUTORS-BADGE:END -->

## Instalação

Efetuar a instalação do [Prettier](https://prettier.io/docs/en/install.html) e em seguida esta extensão.

Para uso local:

```
npm install prettier-plugin-4gl --save-dev
```

Para uso global:

```
npm install prettier-plugin-4gl -g
```

## Opções

### Prettier

Para [detalhes sobre configuração e chaves](https://https://prettier.io/docs/en/options.html) do Prettier.
Abaixo, configurações do _Prettier_ utilizados pela extensão.

| Chave             | Uso                                                                    |
| ----------------- | ---------------------------------------------------------------------- |
| --tab-width <int> | Número de espaços por nível de indentação.                             |
|                   | Padrão: 2                                                              |
| --use-tabs        | Recuar com tabulações em vez de espaços.                               |
|                   | Padrão: false                                                          |
| --insert-pragma   | Insere '@format' no inicio do do arquivo.                              |
|                   | Padrão false                                                           |
| --require-pragma  | Requer que '@prettier' ou '@format' esteja presente no arquivo.        |
|                   | Padrão: false                                                          |
| -w, --write       | Grava o arquivo formato. Cuidado: o arquivo original será sobrescrito. |
|                   | Padrão: false                                                          |

### 4GL

Chaves específicas para formatação de fontes 4GL.

| Chave                             | Uso                                                                        |
| --------------------------------- | -------------------------------------------------------------------------- | ------------- | ----------------------------------------------------------------- |
| --4gl-align-fields                | Alinhar identificação de campos em comandos DEFINE/RECORD. Padrão: false   |
| --4gl-align-comment <int>         | Coluna de alinhamento de comentário de fim de linha. Padrão: 0 (desligado) |
| --4gl-braces                      | Espaçamento entre chaves. Padrão: false                                    |
| --4gl-bracket                     | Espaçamento entre colchetes. Padrão: false                                 |
| --4gl-comma                       | Espaçamento após virgulas de separção. Padrão: false                       |
| --4gl-format-number               | Formata números, p.e. 1234 é formatado para 1,234.                         | Padrão: false |
| --4gl-keywords-case <upper        | lower                                                                      | ignore>       | Coloca palavras-chaves em maiúsculas ou minúsculas. Padrão: upper |
| --4gl-max-empty-line <int>        | Máximo de linhas em branco na sequência. Padrão: 0 (sem limite)            |
| --4gl-math-operators              | Espaçamento em operadores matemáticos. Padrão: false                       |
| --4gl-parenthesis                 | Espaçamento entre parenteses. Padrão: false                                |
| --4gl-string-style <double-quotes | single-quotes                                                              | ignore>       | Usar aspas simples ou duplas em strings. Padrão: ignore           |

### Uso em linha de comando

> Detalhes sobre o [uso em CLI](https://prettier.io/docs/en/cli.html).

> Em determinados sistemas operacionais ou devido a politicas de segurança, pode ser necessário configurações adicionais. Veja a documentação específica do seu sistema operacional ou acione o responsável de infra-estrutura/segurança.

```
npm prettier --parser=4gl [options]
```

### Integração com editores

Ver [Editor Integration](https://prettier.io/docs/en/editors.html).

### Embarcado (API)

Nesse modo, informe as opções usando o nome longo das opções sem o `--`.

```Typescript
  const options: any = { ... } //Prettier Options and Formatter 4GL Options*

let result: any = prettier.format(content, {
    parser: "4gl",
    ...options,
  });
```

## Mantenedor

<table>
  <tr>
    <td align="center"><a href="https://twitter.com/TOTVSDevelopers"><img src="https://avatars2.githubusercontent.com/u/20243897?v=4?s=100" width="100px;" alt=""/><br /><sub><b>TOTVS S.A.</b></sub></a><br /><a href="#maintenance-totvs" title="Maintenance">🚧</a> <a href="#plugin-totvs" title="Plugin/utility libraries">🔌</a> <a href="#projectManagement-totvs" title="Project Management">📆</a></td>
    </tr>
</table>

## Colaboradores

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://github.com/brodao"><img src="https://avatars0.githubusercontent.com/u/949914?v=4?s=50" width="50px;" alt=""/><br /><sub><b>Alan Cândido</b></sub></a><br /><a href="https://github.com/totvs/@totvs/prettier-plugin-4gl/commits?author=brodao" title="Code">💻</a> <a href="https://github.com/totvs/@totvs/prettier-plugin-4gl/commits?author=brodao" title="Documentation">📖</a> <a href="https://github.com/totvs/@totvs/prettier-plugin-4gl/commits?author=brodao" title="Tests">⚠️</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.3](https://github.com/totvs/prettier-plugin-4gl/compare/v0.1.2...v0.1.3) (2021-03-18)

### [0.1.2](https://github.com/totvs/prettier-plugin-4gl/compare/v0.1.1...v0.1.2) (2021-03-17)

### [0.1.1](https://github.com/totvs/prettier-plugin-4gl/compare/v0.0.9...v0.1.1) (2021-03-16)

### [0.0.4](https://github.com/totvs/prettier-plugin-4gl/compare/v0.0.3...v0.0.4) (2020-09-04)


### Bug Fixes

* eliminação da propriedade "bin" ([f388be9](https://github.com/totvs/prettier-plugin-4gl/commit/f388be968ff378d9a6cc9cfe6b469c8f653f7b3b))

### [0.0.9](https://github.com/totvs/prettier-plugin-4gl/compare/v0.0.8...v0.0.9) (2021-02-17)


### Bug Fixes

* ajustes de formatação ([a97f5eb](https://github.com/totvs/prettier-plugin-4gl/commit/a97f5eb1be4943268470b853fc9b78d87021f669))

### [0.0.8](https://github.com/totvs/prettier-plugin-4gl/compare/v0.0.7...v0.0.8) (2021-01-13)

### [0.0.7](https://github.com/totvs/prettier-plugin-4gl/compare/v0.0.6...v0.0.7) (2021-01-13)


### Bug Fixes

* correções no CLI ([e14342c](https://github.com/totvs/prettier-plugin-4gl/commit/e14342c0a136fc68c31289ae413e182f89ce0134))

### [0.0.6](https://github.com/totvs/prettier-plugin-4gl/compare/v0.0.5...v0.0.6) (2021-01-07)


### Bug Fixes

* ajustes de configuração de pacote ([750301d](https://github.com/totvs/prettier-plugin-4gl/commit/750301d418eb023d981e3bd72915d761b87b6df2))

### [0.0.5](https://github.com/totvs/prettier-plugin-4gl/compare/v0.0.3...v0.0.5) (2021-01-07)


### Features

* implementação da opção "alignFields" e correções na formatação ([6ae5f79](https://github.com/totvs/prettier-plugin-4gl/commit/6ae5f796ac2d0d33df2a485bc961b893c95491cd))


### Bug Fixes

* ajustes de formatação ([1895b87](https://github.com/totvs/prettier-plugin-4gl/commit/1895b87e649b7c82e730627942fe5fe3cf54bf8f))

### [0.0.3](https://github.com/totvs/prettier-plugin-4gl/compare/v0.0.2...v0.0.3) (2020-09-04)


### Features

* implementação inicial de CLI ([cbc52de](https://github.com/totvs/prettier-plugin-4gl/commit/cbc52de4916f0e7ff3750ada5e52ffd42995d924))


### Bug Fixes

* ajustes de parametrização ([1f89b91](https://github.com/totvs/prettier-plugin-4gl/commit/1f89b91d3d4dbef0931714cb3641650df866d090))
* correção de depêndencia ([8d60f3f](https://github.com/totvs/prettier-plugin-4gl/commit/8d60f3f239a09ee0a2815771141e0412334d0f6d))
* correção de formatação de diversos blocos e adição de testes ([032891c](https://github.com/totvs/prettier-plugin-4gl/commit/032891cb6a7e82ad4b9f4e4cb03b264cd5579264))
* correção de local de publicação ([051e871](https://github.com/totvs/prettier-plugin-4gl/commit/051e87143762b1cb52f8e91717efe7ba7c451354))
* correção de testes e na gramática ([03fc8bd](https://github.com/totvs/prettier-plugin-4gl/commit/03fc8bdca1a356d4257553adac15db7eebeb4591))
* correção na geração do fonte, testes e configurações ([fdeb970](https://github.com/totvs/prettier-plugin-4gl/commit/fdeb970b46afe9a521e1cc1129c655500075d9ae))
* correção na gramática, procedimentos de testes e efetuado separação de index.js em mais fontes ([82f9047](https://github.com/totvs/prettier-plugin-4gl/commit/82f9047ecba305912fe89ca0b3bd5d1b7e9e77c0))
* correção no processo de geração AST ([ffe4cd4](https://github.com/totvs/prettier-plugin-4gl/commit/ffe4cd48bcff08724ed7eca764588345fff88a1d))
* correção nos processos de formatação ([42bf869](https://github.com/totvs/prettier-plugin-4gl/commit/42bf8697f3db43d52ce65b3236bbb0a9ae0f9ada))
* geração dos fontes formatados ([1de9cb2](https://github.com/totvs/prettier-plugin-4gl/commit/1de9cb2e603848a9f8a1f6ed4fc1b217dbeb336e))
* implementação de mapa de funções para formatação ([8d2b514](https://github.com/totvs/prettier-plugin-4gl/commit/8d2b514a989925738360e634b137cb75125fc946))

### 0.0.2 (2020-08-11)


### Bug Fixes

* correção na estrutura do projeto e configurações ([8999faa](https://github.com/totvs/prettier-plugin-4gl/commit/8999faa8e6427064ab9080572f5fad2e1885f46a))
* correção no arquivo de gramática e retorno de resultado do parser ([25b5576](https://github.com/totvs/prettier-plugin-4gl/commit/25b5576e6ba10706d6efdfaa35c9caf426902f9d))
* correções na configração do projetos ([b3a8d43](https://github.com/totvs/prettier-plugin-4gl/commit/b3a8d434911f9d32547a79a7918ad28a16a414bc))

# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.



src/config.ts
const CATEGORY_4GL: string = '4GL';
const SINCE: string = '0.0.0';

export const options: {} = {
  '4glKeywordsCase': {
    since: SINCE,
    category: CATEGORY_4GL,
    type: 'choice',
    choices: [{ value: 'upper' }, { value: 'lower' }, { value: 'ignore' }],
    default: 'upper',
    description: 'Put keywords to upper or lowser case.',
  },
  '4glStringStyle': {
    since: SINCE,
    category: CATEGORY_4GL,
    type: 'choice',
    choices: [
      { value: 'double-quotes' },
      { value: 'single-quotes' },
      { value: 'ignore' },
    ],
    default: 'ignore',
    description: 'Start/end strings with quotes.',
  },
  '4glFormatNumber': {
    since: SINCE,
    category: CATEGORY_4GL,
    type: 'boolean',
    default: false,
    description: 'Format number, e.g. 1234 is formatted to 1,234.',
  },
  '4glBraces': {
    since: SINCE,
    category: CATEGORY_4GL,
    type: 'boolean',
    default: false,
    description: 'Spacing in Braces operator.',
  },
  '4glBracket': {
    since: SINCE,
    category: CATEGORY_4GL,
    type: 'boolean',
    default: false,
    description: 'Spacing in Bracket operator.',
  },
  '4glParenthesis': {
    since: SINCE,
    category: CATEGORY_4GL,
    type: 'boolean',
    default: false,
    description: 'Spacing in Parenthesis operator.',
  },
  '4glComma': {
    since: SINCE,
    category: CATEGORY_4GL,
    type: 'boolean',
    default: false,
    description: 'Spacing in Comma operator.',
  },
  '4glMathOperators': {
    since: SINCE,
    category: CATEGORY_4GL,
    type: 'boolean',
    default: false,
    description: 'Spacing in Mathematical operators.',
  },
  // '4glAlignFields': {
  //   since: SINCE,
  //   category: CATEGORY_4GL,
  //   type: 'int',
  //   default: 0,
  //   description:
  //     'Align field name with padding spaces in DEFINE/RECORD statment.',
  // },
  // '4glAlignComment': {
  //   since: SINCE,
  //   category: CATEGORY_4GL,
  //   type: 'boolean',
  //   default: false,
  //   description: 'End of line comment alignment column.',
  // },
  '4glMaxEmptyLines': {
    since: SINCE,
    category: CATEGORY_4GL,
    type: 'int',
    default: 0,
    description: 'Maximum blank lines in sequence (0, no limits).',
  },
};

export const prettierOptions: any = {
  printWidth: 80,
  useTabs: false,
  tabWidth: 2,
  insertPragma: true,
  requirePragma: false,
};
